### PR TITLE
Added IFileSystemWatcher, IFile and IDirectory interfaces

### PR DESCRIPTION
**Pull request by wexman**
I have added the IFileSystemWatcher, IFile and IDirectory interface for easier mocking...

### CODE REVIEW

**Comment by fgreinacher** [comment]
This would implement #175, though I'm not sure if a breaking API change like that would produce more negative than positive echo. 

@jpreese @robkeim What's your opinion?

**Comment by wexman** [comment]
That might indeed be a breaking change, but a rather theoretical one, meaning it won't actually break because the interfaces have the same members as the abstract classes...

File: System.IO.Abstractions/IDirectory.cs [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        void SetLastAccessTime(string path, DateTime lastAccessTime);
+        void SetLastAccessTimeUtc(string path, DateTime lastAccessTimeUtc);
+        void SetLastWriteTime(string path, DateTime lastWriteTime);
+        void SetLastWriteTimeUtc(string path, DateTime lastWriteTimeUtc);
+
+#if NET40
+        DirectoryInfoBase CreateDirectory(string path, DirectorySecurity directorySecurity);
+        string[] GetLogicalDrives();
+#endif
+
```
Comment (robkeim) [inline System.IO.Abstractions/IDirectory.cs]: Minor: extra newline here

**Comment by robkeim** [comment]
I also agree with the direction that #175 wants to go, how difficult do you think the breaking changes would be for people consuming the library to fix in their solutions @fgreinacher?

If you're using the `var` keyword the change shouldn't be breaking for local variables inside functions:
```
var fileSystemWatcherFactory = new FileSystemWatcherFactory();
var fileSystemWatcher = fileSystemWatcherFactory.CreateNew(); // No breaking change here changing from FileSystemWatcherBase to IFileSystemWatcher
```

For instances where there are breaking changes (explicitly typed local variables, function parameters, function return type, etc, is a simple replace of `XyzBase` to `IXyz` sufficient for all the types that introduced the interface? That's definitely a naive approach, but I can't think of situations off the top of my head where doing so wouldn't be sufficient or would introduce other issues.

If that's the case then I don't think we should be too worried about this breaking change.

**Comment by wexman** [comment]
BTW, do you have any idea why the comments from my previous pull request show up here? Did i do something wrong?

**Comment by fgreinacher** [comment]
> BTW, do you have any idea why the comments from my previous pull request show up here? Did i do something wrong?

(I assume comments == commits) That's just happening because we are updating your branch from `master`. It's fine :)

**Comment by fgreinacher** [comment]
OK - let's do this. A couple of things that I'd ask you to do:

1. Update the version in `version.json` to 4.0 (as this is a breaking change)
1. Add `<inheritdoc>`comments to the new interfaces (just mirror what's done in the implementations) so that we get proper XML docs.
1. Think about whether it makes sense to add a short chapter to `README.md` that describes the new use case this change enables

**Comment by wexman** [comment]
Oh, ok. But if we're going that route anyway, maybe we should do the same for all other cases as well? For example, IPath instead of PathBase, IFileInfo instead of FileInfoBase, IDirectoryInfo instead of DirectoryInfoBase and so on?

**Comment by robkeim** [comment]
@wexman Yes I think it definitely makes sense to have a consistent API

**Comment by wexman** [comment]
Ok, it's done...

**Comment by fgreinacher** [comment]
Merged with some adaptions.  Thanks a lot for your work @wexman!

**Pull request merged**